crates/rspack_regex/src/algo.rs

```diff
@@ -0,0 +1,99 @@
+use std::sync::Arc;
+
+use regex_syntax::hir::literal::ExtractKind;
+use regex_syntax::hir::{Hir, HirKind, Look};
+use rspack_error::{internal_error, Error};
+
+#[derive(Clone)]
+pub enum Algo {
+  FastCustom(Arc<dyn Fn(&str) -> bool + Send + Sync>),
+  Regress(regress::Regex),
+}
+
+impl std::fmt::Debug for Algo {
+  fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+    match self {
+      Self::FastCustom(_) => write!(f, "FastCustom(...)"),
+      Self::Regress(arg0) => f.debug_tuple("Regress").field(arg0).finish(),
+    }
+  }
+}
+
+impl Algo {
+  pub fn new(expr: &str, flags: &str) -> Result<Algo, Error> {
```

Comment:
It would be great if we provide some documentation for why we are doing this.

Reply:
Algo is the abbreviation of `Algorithm` , this layer control how did `RspackRegex` match a string literal, fast  string search, or fallback to a normal regress regex matching.